flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/InputGateWithMetrics.java

```diff
@@ -0,0 +1,119 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskmanager;
+
+import org.apache.flink.metrics.Counter;
+import org.apache.flink.runtime.event.TaskEvent;
+import org.apache.flink.runtime.io.network.partition.consumer.BufferOrEvent;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGateListener;
+import org.apache.flink.runtime.metrics.groups.TaskIOMetricGroup;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * The InputGateWithMetrics is wrapped in task with {@link InputGate} provided by
+ * shuffle service. It is mainly used for increasing general input metrics from
+ * {@link TaskIOMetricGroup}.
+ */
+public class InputGateWithMetrics implements InputGate, InputGateListener {
+
+	private final InputGate inputGate;
+
+	private final Counter numBytesInCounter;
+
+	/** Registered listener to forward buffer notifications to. */
+	private volatile InputGateListener inputGateListener;
+
+	InputGateWithMetrics(InputGate inputGate, Counter numBytesInCounter) {
+		this.inputGate = checkNotNull(inputGate);
+		this.numBytesInCounter = checkNotNull(numBytesInCounter);
+	}
+
+	@Override
+	public int getNumberOfInputChannels() {
+		return inputGate.getNumberOfInputChannels();
+	}
+
+	@Override
+	public String getOwningTaskName() {
+		return inputGate.getOwningTaskName();
+	}
+
+	@Override
+	public boolean isFinished() {
+		return inputGate.isFinished();
+	}
+
+	@Override
+	public void requestPartitions() throws IOException, InterruptedException {
+		inputGate.requestPartitions();
+	}
+
+	@Override
+	public Optional<BufferOrEvent> getNextBufferOrEvent() throws IOException, InterruptedException {
+		Optional<BufferOrEvent> bufferOrEvent = inputGate.getNextBufferOrEvent();
+		if (bufferOrEvent.isPresent()) {
+			numBytesInCounter.inc(bufferOrEvent.get().getSize());
+		}
+
+		return bufferOrEvent;
+	}
+
+	@Override
+	public Optional<BufferOrEvent> pollNextBufferOrEvent() throws IOException, InterruptedException {
+		return inputGate.pollNextBufferOrEvent();
+	}
+
+	@Override
+	public void sendTaskEvent(TaskEvent event) throws IOException {
+		inputGate.sendTaskEvent(event);
+	}
+
+	@Override
+	public void registerListener(InputGateListener inputGateListener) {
+		if (this.inputGateListener == null) {
+			this.inputGateListener = inputGateListener;
+		} else {
+			throw new IllegalStateException("Multiple listeners");
+		}
+
+		inputGate.registerListener(this);
+	}
+
+	@Override
+	public int getPageSize() {
+		return inputGate.getPageSize();
+	}
+
+	@Override
+	public void close() throws Exception {
```

Comment:
should we not forward `close` to `inputGate` as well?

Reply:
In task class we would finally maintain only the array of `InputGateWithMetrics` finally, so the close called from task would focus on `InputGateWithMetrics`. Otherwise the task would maintain two arrays `SingleInputGate` and `InputGateWithMetrics` separately.

Reply:
If we change to wrap the `InputGateWithMetrics` in the constructor as you mentioned below, we should forward `close` to `InputGate` as well.